src/main/frontend/app/routes/editor/editor.tsx

```diff
@@ -30,6 +45,149 @@ export default function CodeEditor() {
+
+      // apply highlight decoration
+      decorationIdsReference.current = editor.deltaDecorations(decorationIdsReference.current, [
+        {
+          range: { startLineNumber: lineNumber, startColumn: 1, endLineNumber: lineNumber, endColumn: 1 },
+          options: {
+            isWholeLine: true,
+            className: 'highlight-line',
+          },
+        },
+      ])
+
+      // remove highlight after 2s
+      const t = setTimeout(() => {
+        decorationIdsReference.current = editor.deltaDecorations(decorationIdsReference.current, [])
+      }, 2000)
+
+      // optional cleanup if component unmounts before timeout
+      return () => clearTimeout(t)
+    }, 50)
+  }, [xmlContent, activeTab])
+
+  useEffect(() => {
+    // Handles all the suggestions
+    if (!editorReference.current) return
+    const monacoInstance = (globalThis as any).monaco
+    if (!monacoInstance) return
+
+    // Keep latest elements in a ref so provider callbacks always see current data
+    const elementsReference = { current: elements }
+
+    // Element suggestions
+    const elementProvider = monacoInstance.languages.registerCompletionItemProvider('xml', {
+      triggerCharacters: ['<'],
+      provideCompletionItems: () => {
+        return {
+          suggestions: Object.values(elementsReference.current).map((element: any) => {
+            // Mandatory attributes
+            const mandatoryAttributes = Object.entries(element.attributes || {})
+              .filter(([_, attribute]) => attribute.mandatory)
+              .map(([name]) => `${name}="\${${name}}"`)
+
+            // Snippet for tag + mandatory attributes
+            // eslint-disable-next-line sonarjs/no-nested-template-literals
+            const insertText = `${element.name}${mandatoryAttributes.length > 0 ? ` ${mandatoryAttributes.join(' ')}` : ''}>$0</${element.name}`
+
+            return {
+              label: element.name,
+              kind: monacoInstance.languages.CompletionItemKind.Class,
+              insertText,
+              insertTextRules: monacoInstance.languages.CompletionItemInsertTextRule.InsertAsSnippet,
+              documentation: element.description || '',
+            }
+          }),
+        }
+      },
+    })
+
+    // Attribute suggestions
+    const attributeProvider = monacoInstance.languages.registerCompletionItemProvider('xml', {
+      triggerCharacters: [' '],
+      provideCompletionItems: (model, position) => {
+        const line = model.getLineContent(position.lineNumber)
+        const textBeforeCursor = line.slice(0, position.column - 1)
+
+        // Don't show suggestions if cursor is inside quotes
+        const quotesBefore = (textBeforeCursor.match(/"/g) || []).length
+        if (quotesBefore % 2 === 1) {
+          // Odd number of quotes -> cursor is inside an attribute value
+          return { suggestions: [] }
+        }
+
+        const tagMatch = line.slice(0, position.column - 1).match(/<(\w+)/)
```

Comment:
Stuff like this would be nice to encapsulate as well. When reading "tagMatchesOnLine" you don't have to think twice about it.